[iOS] Element Fullscreen does not respect viewport-fit

jernoble opened the pull request:
#### 56d49b081448ceac6d4f2fbd296ffe4e80b15aa6
<pre>
[iOS] Element Fullscreen does not respect viewport-fit
<a href="https://bugs.webkit.org/show_bug.cgi?id=264012">https://bugs.webkit.org/show_bug.cgi?id=264012</a>
<a href="https://rdar.apple.com/117304719">rdar://117304719</a>

Reviewed by Wenson Hsieh and Tim Horton.

Tests: fast/viewport/ios/full-screen-safe-area-insets.html
       fullscreen/full-screen-document-background-color.html

When configuring the WKWebView during the enter fullscreen operation, various settings of the view
must be returned to their default state for the &quot;automatic&quot; avoid-safe-areas behavior to kick in.
For some calls made by clients, there is no way to reset those behaviors to default, and the
existing implementation merely overrode those settings with other non-default values. The end
result was that all fullscreen content was behaving as if `viewport-fit=cover` was specified, which
allowed some content to slip into the safe areas.

Additionally, when embedded content is taken fullscreen, the viewport settings of that embedded
iframe are not respected, and the embedded content uses the viewport settings of whatever page
embedded it. Also, the fullscreen element&apos;s background is not used in the overflow areas when
iframe content is in fullscreen.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::updateViewportArguments):
* Source/WebCore/dom/FullscreenManager.cpp:
(WebCore::FullscreenManager::dispatchFullscreenChangeOrErrorEvent):
(WebCore::FullscreenManager::deepestFullscreenDocument const):
* Source/WebCore/dom/FullscreenManager.h:
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::documentBackgroundColor const):
* Source/WebCore/page/Page.cpp:
(WebCore::viewportDocumentForFrame):
(WebCore::Page::viewportArguments const):
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _resetScrollViewInsetAdjustmentBehavior]):
(-[WKWebView _haveSetUnobscuredSafeAreaInsets]):
(-[WKWebView _resetUnobscuredSafeAreaInsets]):
(-[WKWebView _hasOverriddenLayoutParameters]):
(-[WKWebView _viewLayoutSizeOverride]):
(-[WKWebView _minimumUnobscuredSizeOverride]):
(-[WKWebView _maximumUnobscuredSizeOverride]):
(-[WKWebView _resetObscuredInsets]):
(-[WKWebView _clearOverrideLayoutParameters]):
* Source/WebKit/UIProcess/ios/WKContentView.mm:
(-[WKContentView setFrame:]):
* Source/WebKit/UIProcess/ios/WKScrollView.h:
* Source/WebKit/UIProcess/ios/WKScrollView.mm:
(-[WKScrollView _contentInsetWasExternallyOverridden]):
(-[WKScrollView _resetContentInset]):
(-[WKScrollView _resetContentInsetAdjustmentBehavior]):
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.mm:
(-[WKFullScreenViewController viewDidLayoutSubviews]):
(-[WKFullScreenViewController viewWillTransitionToSize:withTransitionCoordinator:]):
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
(WebKit::WKWebViewState::applyTo):
(WebKit::WKWebViewState::store):
(-[WKFullScreenWindowController enterFullScreen:]):
(-[WKFullScreenWindowController beganEnterFullScreenWithInitialFrame:finalFrame:]):

Canonical link: <a href="https://commits.webkit.org/270199@main">https://commits.webkit.org/270199@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1f72f3883cd989428a30fb2fa23ee39238757cad

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/24816 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/3361 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/26070 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/26935 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/22782 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/25085 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/5040 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/798 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/23113 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/25061 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/2404 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/21427 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/27518 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/2111 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/22358 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/28512 "Found 1 new test failure: imported/w3c/web-platform-tests/webcodecs/audio-encoder-codec-specific.https.any.html (failure)") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/22634 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/22704 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/26330 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2049 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/354 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/3338 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/22108 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5947 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/2499 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/2400 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jernoble commented on Source/WebKit/UIProcess/ios/WKContentView.mm:
```diff
@@ -427,6 +427,11 @@ - (void)didMoveToWindow
+{
+    [super setFrame:frame];
+}
+
```
Accidental debugging aid left in; will remove.

hortont424 commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -3409,6 +3417,46 @@ - (void)_didRemoveLayerForFindOverlay
+
+- (CGSize)_viewLayoutSizeOverride
+{
+    return _viewLayoutSizeOverride.value_or(CGSizeMake(0, 0));
```
Why are these getting flattened to something that is indistinguishable from a valid overridden value? Can you not just return an optional?

jernoble replied:
I didn't want to add a C++ optional to an Objective-C header.

jernoble replied:
But also, that's why there's a getter for whether any values are overridden.

jernoble replied:
@hortont424 pointed out that there are already C++ concepts in that header; will fix.

smfr commented on LayoutTests/fast/viewport/ios/full-screen-safe-area-insets.html:
```diff
@@ -0,0 +1,130 @@
+        run('frame.src = "resources/viewport-fit-cover.html"');
+        await waitFor(frame, 'load');
+
+        await run('enterFullscreen()');
```
I thought we couldn't test fullscreen in layout tests?

jernoble replied:
We can't test fullscreen in API tests; WKTR/DRT do a "fake" fullscreen which doesn't involve anything at the UI level.

smfr commented on LayoutTests/fast/viewport/ios/resources/viewport-fit-contain.html:
```diff
@@ -0,0 +1,15 @@
+<!DOCTYPE html>
+<html>
+<head>
+	<meta name="viewport" content="width=device-width, initial-scale=1,viewport-fit=contain">
```
Tab here.

smfr commented on LayoutTests/fast/viewport/ios/resources/viewport-fit-cover.html:
```diff
@@ -0,0 +1,15 @@
+<!DOCTYPE html>
+<html>
+<head>
+	<meta name="viewport" content="width=device-width, initial-scale=1,viewport-fit=cover">
```
And here.

smfr commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -774,6 +777,26 @@ bool FullscreenManager::isSimpleFullscreenDocument() const
+    if (!fullscreenElement)
+        return nullptr;
+
+    if (!is<HTMLFrameOwnerElement>(fullscreenElement))
```
Use `dynamicDowncast<>` rather than `is<>` followed by `downcast<>`

hortont424 reviewed (approved):
This needs review from as many people as you can muster :P

smfr commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -774,6 +777,26 @@ bool FullscreenManager::isSimpleFullscreenDocument() const
+    if (!contentDocument)
+        return m_document.ptr();
+
+    auto* fullscreenDocument = contentDocument->fullscreenManager().deepestFullscreenDocument();
```
"deepest" seems like the wrong description. "deepest" implies "leaf-most" but I think this is "outer fullscreen document"?

smfr replied:
Also, since this crosses frame boundaries, and it seems like `FullscreenManager` is a per-Document thing, this function should live on Page, and may have to be different in future for site isolation.

jernoble replied:
Can't just be `currentFullscreenDocument()` because multiple documents are in fullscreen simultaneously (specifically all the ancestor documents of the "outermost" one).

Can't just be `topmostFullscreenDocument()` because that's too close to `topDocument()` and implies a heirarchy that goes in a different direction.

So, I'd be happy to change this to `outermostFullscreenDocument()`. And I'll move it into Page.

smfr commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -4641,19 +4641,35 @@ Color LocalFrameView::documentBackgroundColor() const
-    auto* htmlElement = m_frame->document()->documentElement();
-    auto* bodyElement = m_frame->document()->bodyOrFrameset();
+    if (auto* fullscreenDocument = document->fullscreenManager().deepestFullscreenDocument())
+        document = fullscreenDocument;
```
It makes me very nervous to reassign `document` to a document from another, potentially cross-origin frame. Also, with site isolation, there is no way this can work. I think this code has to be different.

jernoble replied:
The caller of LocalFrameView::documentBackgroundColor() is going to have to call something like `outermostFullscreenView()->documentBackgroundColor()` rather than `mainFrameView()->documentBackgroundColor()`. In which case, this existing code could work without modification.

But for the `document` reassignment, I renamed it `backgroundDocument` and moved its initializer into a lambda to hopefully clarify that no documents are being re-assigned here.

smfr commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
```diff
@@ -201,6 +202,14 @@ enum class TapHandlingResult : uint8_t;
 
+@property (nonatomic, readonly) BOOL _haveSetUnobscuredSafeAreaInsets;
+@property (nonatomic, readonly) BOOL _hasOverriddenLayoutParameters;
+@property (nonatomic, readonly) std::optional<CGSize> _viewLayoutSizeOverride;
```
Using C++ in an Objective-C header.

jernoble replied:
This header already has C++ in it.

smfr commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -140,54 +141,106 @@ static void resizeScene(UIWindowScene *scene, CGSize size, CompletionHandler<voi
     CGFloat _savedViewScale = 1.0;
     CGFloat _savedZoomScale = 1;
+    CGFloat _savedContentZoomScale = 1;
+    BOOL _savedContentInsetWasExternallyOverridden;
```
Why is this one not initialized?

jernoble replied:
Good catch. Will fix.

whsieh commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -140,54 +141,106 @@ static void resizeScene(UIWindowScene *scene, CGSize size, CompletionHandler<voi
-        _savedObscuredInsets = [webView _obscuredInsets];
-        _savedEdgeInset = [[webView scrollView] contentInset];
-        _savedContentOffset = [[webView scrollView] contentOffset];
+        auto* scrollView = (WKScrollView *)[webView scrollView];
```
```suggestion
        auto *scrollView = [webView _wkScrollView];
```

Nit - there's an internal helper that already grabs the scroll view as a `WKScrollView`.

nt1m commented on Source/WebCore/page/Page.cpp:
```diff
@@ -3794,6 +3815,33 @@ void Page::setFullscreenControlsHidden(bool hidden)
+Document* Page::outermostFullscreenDocument() const
+{
+#if ENABLE(FULLSCREEN_API)
+    auto* localMainFrame = dynamicDowncast<LocalFrame>(m_mainFrame.get());
```
Does anything in the method require smart pointers?

jernoble replied:
As a `const` method, it should _in theory_ not mutate anything in such a way that `m_mainFrame` changes during the runtime of the method. In practice, calling callbacks and dispatching events make that not true, but neither of those things happens here.

jernoble replied:
I'll modify all these raw pointers to use CheckedPtr however.

nt1m commented on Source/WebCore/page/Page.cpp:
```diff
@@ -3794,6 +3815,33 @@ void Page::setFullscreenControlsHidden(bool hidden)
 #endif
 }
 
+Document* Page::outermostFullscreenDocument() const
```
Not sure outermost is how I would describe it. If anything, it feels more inner than outer.    `deepest` described it better IMO.

jernoble replied:
Happy to paint this bikeshed whatever color people agree upon.

cdumez commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -680,6 +680,9 @@ void FullscreenManager::dispatchFullscreenChangeOrErrorEvent(Deque<GCReachableRe
         }
 
+        // Gaining or losing fullscreen state may change viewport arguments
+        node->document().updateViewportArguments();
```
`node->protectedDocument()->updateViewportArguments();` per our new smart pointers idioms

jernoble committed:
[iOS] Element Fullscreen does not respect viewport-fit
https://bugs.webkit.org/show_bug.cgi?id=264012
rdar://117304719

Reviewed by Wenson Hsieh and Tim Horton.

Tests: fast/viewport/ios/full-screen-safe-area-insets.html
       fullscreen/full-screen-document-background-color.html

When configuring the WKWebView during the enter fullscreen operation, various settings of the view
must be returned to their default state for the "automatic" avoid-safe-areas behavior to kick in.
For some calls made by clients, there is no way to reset those behaviors to default, and the
existing implementation merely overrode those settings with other non-default values. The end
result was that all fullscreen content was behaving as if `viewport-fit=cover` was specified, which
allowed some content to slip into the safe areas.

Additionally, when embedded content is taken fullscreen, the viewport settings of that embedded
iframe are not respected, and the embedded content uses the viewport settings of whatever page
embedded it. Also, the fullscreen element's background is not used in the overflow areas when
iframe content is in fullscreen.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::updateViewportArguments):
* Source/WebCore/dom/FullscreenManager.cpp:
(WebCore::FullscreenManager::dispatchFullscreenChangeOrErrorEvent):
(WebCore::FullscreenManager::deepestFullscreenDocument const):
* Source/WebCore/dom/FullscreenManager.h:
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::documentBackgroundColor const):
* Source/WebCore/page/Page.cpp:
(WebCore::viewportDocumentForFrame):
(WebCore::Page::viewportArguments const):
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _resetScrollViewInsetAdjustmentBehavior]):
(-[WKWebView _haveSetUnobscuredSafeAreaInsets]):
(-[WKWebView _resetUnobscuredSafeAreaInsets]):
(-[WKWebView _hasOverriddenLayoutParameters]):
(-[WKWebView _viewLayoutSizeOverride]):
(-[WKWebView _minimumUnobscuredSizeOverride]):
(-[WKWebView _maximumUnobscuredSizeOverride]):
(-[WKWebView _resetObscuredInsets]):
(-[WKWebView _clearOverrideLayoutParameters]):
* Source/WebKit/UIProcess/ios/WKContentView.mm:
(-[WKContentView setFrame:]):
* Source/WebKit/UIProcess/ios/WKScrollView.h:
* Source/WebKit/UIProcess/ios/WKScrollView.mm:
(-[WKScrollView _contentInsetWasExternallyOverridden]):
(-[WKScrollView _resetContentInset]):
(-[WKScrollView _resetContentInsetAdjustmentBehavior]):
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.mm:
(-[WKFullScreenViewController viewDidLayoutSubviews]):
(-[WKFullScreenViewController viewWillTransitionToSize:withTransitionCoordinator:]):
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
(WebKit::WKWebViewState::applyTo):
(WebKit::WKWebViewState::store):
(-[WKFullScreenWindowController enterFullScreen:]):
(-[WKFullScreenWindowController beganEnterFullScreenWithInitialFrame:finalFrame:]):

Canonical link: https://commits.webkit.org/270199@main
